java-checks/src/main/java/org/sonar/java/checks/synchronization/DoubleCheckedLockingCheck.java

```diff
@@ -0,0 +1,233 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.synchronization;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import javax.annotation.CheckForNull;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+import static org.sonar.plugins.java.api.tree.Tree.Kind.EQUAL_TO;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.IDENTIFIER;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.IF_STATEMENT;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.MEMBER_SELECT;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.NULL_LITERAL;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.SYNCHRONIZED_STATEMENT;
+
+@Rule(key = "S2168")
+public class DoubleCheckedLockingCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MESSAGE = "Remove this dangerous instance of double-checked locking.";
+
+  private Deque<ExpressionTree> ifConditionStack = new LinkedList<>();
+  private Deque<Tree> synchronizedStmtStack = new LinkedList<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(IF_STATEMENT, SYNCHRONIZED_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(IF_STATEMENT)) {
+      visitIfStatement((IfStatementTree) tree);
+    }
+    if (tree.is(SYNCHRONIZED_STATEMENT)) {
+      synchronizedStmtStack.add(tree);
+    }
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (tree.is(IF_STATEMENT) && fieldNotInitializedTest(((IfStatementTree) tree).condition())) {
+      ifConditionStack.pollLast();
+    }
+    if (tree.is(SYNCHRONIZED_STATEMENT)) {
+      synchronizedStmtStack.pollLast();
+    }
+  }
+
+  private void visitIfStatement(IfStatementTree ifTree) {
+    if (!fieldNotInitializedTest(ifTree.condition())) {
+      return;
+    }
+    ifConditionStack.add(ifTree.condition());
+    if (!insideCriticalSection() || !isNestedIfStatement()) {
+      return;
+    }
+    patternDetected(ifTree);
+  }
+
+  private boolean isNestedIfStatement() {
+    return ifConditionStack.size() > 1;
```

Comment:
I think this is premature optimization, the stack will not be very deep usually. Also notice that I am checking for `> 1` , not `> 0` , so the semantic is OK